### DISCUSSION TITLE
Understanding and fixing the "Cannot resolve a Relative URI Reference since there's no in-scope Base URI!" Error

### DISCUSSION THREAD

**Discussion by bcalco** [dotnetrdf/dotnetrdf#438] [Q&A]
I am getting a "Cannot resolve a Relative URI Reference since there is no in-scope Base URI!" message when running the following SPARQL query (from the test suite for "Learning SPARQL"):

```sparql
# filename: ex007.rq

PREFIX ab: <http://learningsparql.com/ns/addressbook#> 

SELECT ?craigEmail FROM <ex002.ttl>
WHERE
{ ab:craig ab:email ?craigEmail . }
```

Expectation is that the query processor will work out that ex002.ttl is in the same folder from which the query is being called and load it into the graph before processing.

I'm noticing that `arq` from the Jena project handles this fine without feeding it the data file via the `--data` switch when only the query file (above) is specified. When my application calls the same query, that error is produced whether or not I provide the data file explicitly in the query and load it before executing the query via the Leviathan processor. Here are the contents of ex002.ttl, for reference:

```ttl
# filename: ex002.ttl

@prefix ab: <http://learningsparql.com/ns/addressbook#> .

ab:richard ab:homeTel "(229) 276-5135" . 
ab:richard ab:email   "richard49@hotmail.com" . 

ab:cindy ab:homeTel "(245) 646-5488" . 
ab:cindy ab:email   "cindym@gmail.com" . 

ab:craig ab:homeTel "(194) 966-1505" . 
ab:craig ab:email   "craigellis@yahoo.com" . 
ab:craig ab:email   "c.ellis@usairwaysgroup.com" .
```

I gues the problem is with the `FROM` clause's use of<...> referring to the local filesystem? I don't want to have to force the user to specify the base URI from the query if they don't have to, so behaving like `arq` here is important. How should I solve this problem?

Also, minor question which I'll probably figure out 5 minutes after asking: How do I ensure the results use prefixes or `:` for the base prefix instead of full URIs?

**Comment by bcalco** [comment]
It looks like the error comes from here: https://github.com/dotnetrdf/dotnetrdf/blob/master/Libraries/dotNetRDF/Core/Tools.cs

at line 165 where  and RdfParseException is being thrown:

```csharp
           // if (!baseUri.Equals(String.Empty))
           // {
           // ...
           }
            else
            {
                if (uriref.Equals(String.Empty))
                {
                    throw new RdfParseException("Cannot use an Empty URI to refer to the document Base URI since there is no in-scope Base URI!");
                }

                try
                {
                    return new Uri(FixMalformedUriStrings(uriref), UriKind.Absolute).AbsoluteUri;
                }
                catch (UriFormatException)
                {
                    throw new RdfParseException("Cannot resolve a Relative URI Reference since there is no in-scope Base URI!");
                }
            }
        }

```

So it looks like the `FROM <...>` clause is failing to parse. Is there an alternate syntax supported by standard SPARQL for this that will bake both dotNetRDF and Jena happy?

**Reply by bcalco** [reply]
It seems I need to set `parser.DefaultBaseUri` to something indicating the local filesystem? But I will not always want to do that. Some of the queries tested like this refer to http endpoints, as well. How should I handle this?

**Comment by bcalco** [comment]
OK So I got rid of the exception by setting the `parser.DefaultBaseUri` to `new Uri("file://" + Directory.CurrentDirectory() + "/"` however the data coming back is not correct (i.e., there is no data):

```
+--------------+
| ?craigEmail  |
+==============+
+--------------+
```

`arq` produces:

```
--------------------------------
| craigEmail                   |
================================
| "craigellis@yahoo.com"       |
| "c.ellis@usairwaysgroup.com" |
--------------------------------
```

So I am needing some advice how to get the query processor to load the file from the URI specified, which given the default base URI, should resolve the the correct file referenced.

**Reply by bcalco** [reply]
I tried to understand what's going on by using the `ExplainQueryProcessor`, but all it added was

```

  [Thread ID 1] Depth 1 VDS.RDF.Query.Algebra.Select Start
  [Thread ID 1] Depth 2 VDS.RDF.Query.Algebra.Bgp Start
  [Thread ID 1] Depth 1 VDS.RDF.Query.Algebra.Bgp End
  [Thread ID 1] Depth 0 VDS.RDF.Query.Algebra.Select End
+--------------+
| ?craigEmail  |
+==============+
+--------------+
```

I'll tweak information level but... not especially informative out of the box.

**Reply by bcalco** [reply]
So, ExplanationLevel.Full was a bit more informative but I still don't know why it's failing to give correct results:

```
Z:\Documents\GitHub\Apex.Unify\Tests\sparql>adb query --explain --query ex007.rq
  Switching to named graph as merge of the following graphs:
  Switching to named graph as merge of the following graphs:
  [Thread ID 1] Depth 1 VDS.RDF.Query.Algebra.Select Start
  [Thread ID 1] Depth 1 VDS.RDF.Query.Algebra.Select Start
  [Thread ID 1] Depth 2 VDS.RDF.Query.Algebra.Bgp Start
  [Thread ID 1] Depth 2 VDS.RDF.Query.Algebra.Bgp Start
  <http://learningsparql.com/ns/addressbook#craig> <http://learningsparql.com/ns/addressbook#email> ?craigEmail .
  <http://learningsparql.com/ns/addressbook#craig> <http://learningsparql.com/ns/addressbook#email> ?craigEmail .
  [Thread ID 1] Depth 2 VDS.RDF.Query.Algebra.Bgp Took 00:00:00.0040319
  [Thread ID 1] Depth 2 VDS.RDF.Query.Algebra.Bgp Took 00:00:00.0040319
  [Thread ID 1] Depth 2 VDS.RDF.Query.Algebra.Bgp 1 Results(s)
  [Thread ID 1] Depth 2 VDS.RDF.Query.Algebra.Bgp 1 Results(s)
  [Thread ID 1] Depth 1 VDS.RDF.Query.Algebra.Bgp End
  [Thread ID 1] Depth 1 VDS.RDF.Query.Algebra.Bgp End
  [Thread ID 1] Depth 1 VDS.RDF.Query.Algebra.Select Took 00:00:00.0292463
  [Thread ID 1] Depth 1 VDS.RDF.Query.Algebra.Select Took 00:00:00.0292463
  [Thread ID 1] Depth 1 VDS.RDF.Query.Algebra.Select 1 Results(s)
  [Thread ID 1] Depth 1 VDS.RDF.Query.Algebra.Select 1 Results(s)
  [Thread ID 1] Depth 0 VDS.RDF.Query.Algebra.Select End
  [Thread ID 1] Depth 0 VDS.RDF.Query.Algebra.Select End
+--------------+
| ?craigEmail  |
+==============+
|              |
+--------------+
```

It _looks_ like it found 1 result, twice (there are 2 expected results) but I'm not seeing them. I *am* seeing an empty string result (previously there were no rows). So I'm not sure why it's not finding the bound value(s).

**Comment by bcalco** [comment]
OK I found [the discussion of FROM and FROM NAMED in the documentation](https://dotnetrdf.org/docs/stable/user_guide/Querying-With-SPARQL.html#from-and-from-named). I really do need, as much as possible, for those references to resolve whether from disk or the web, so I'm studying the `DiskDemandTripleStore` and `WebDemandTripleStore` for clues about implementing one that can do both.

**Reply by bcalco** [reply]
However, I would like someone to explain the explain query results to me. It seems to find at least one result that is not displayed and I'm curious in general to fathom the explain query processor better. It doesn't seem to be particularly discussed in the documentation and had a lot of options that I'd like to understand better.

**Reply by kal** [reply]
You are correct, you will need to use a `DiskDemandTripleStore` or `WebDemandTripleStore`. The `WebDemandTripleStore` will resolve local `file:` URIs so that is the way to get a store that will do what you want. 

When parsing the SPARQL query you need to provide the query parser with the default base URI to use for resolving relative URI references. It is arguably a bug (or maybe best to say a "missing feature") that the parser doesn't set this to the file URI if you use the `LoadFromFile` method.

The following code works for me with a local copy of the query and data files. Note that the `DefaultBaseUri` property is set using the file: URI of the query file being parsed.

```c#
            var queryPath = Path.GetFullPath(Path.Combine("resources", "sparql", "dataset-1.rq"));
            var queryParser = new SparqlQueryParser { DefaultBaseUri = new Uri(queryPath) };
            var q = queryParser.ParseFromFile(queryPath);
            var store = new WebDemandTripleStore();
            var queryProcessor = new LeviathanQueryProcessor(store);
            var result = q.Process(queryProcessor);
            SparqlResultSet resultSet = Assert.IsAssignableFrom<SparqlResultSet>(result);
            Assert.Equal(2, resultSet.Count);
```

With regards to the Explain output. I think what you are seeing in the trace is a single null-binding result - i.e. ?craigEmail is null.

**Reply by bcalco** [reply]
Thanks @kal, `WebDemandTripleStore` did the trick. My next question in processing results is how to make the output optionally less verbose for display in grids and on the console. By that I mean, using qnames instead of fully qualified URIs, and removing type annotations when displaying the results. I'd like a single knob, or somehow to create one if it doesn't exist, that can tune those results, mainly based on whether I expect a human or a machine to process the results.
